### PR TITLE
fix: update the logic of reconstruction in RLT

### PR DESCRIPTION
**Pull request by tiny-xie**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/modules/rlt_token_transformer.py [hunk 1]
```diff
@@ -225,55 +223,88 @@ def __init__(
...
-        self.output_proj = (
-            nn.Linear(self.embed_dim, self.input_dim)
-            if self.input_dim != self.embed_dim
-            else nn.Identity()
-        )
+        self.output_proj = nn.Linear(self.embed_dim, self.input_dim)
 
-    def forward(self, rl_tokens: torch.Tensor, target_seq_len: int | None = None):
-        target_seq_len = int(target_seq_len or self.prefix_seq_len)
+    def forward(
```
Comment (F9rozen) [inline rlinf/models/embodiment/modules/rlt_token_transformer.py:228]: `RLTTokenDecoder.forward` is now a teacher-forced AR decoder and changes the public `decode(...)` contract (needs `target_embeddings` instead of `target_seq_len`).

This is incompatible with existing stage-1 RLT checkpoints (removed `prefix_token_embed` / old decoder PE, and `output_proj` is always `nn.Linear` even when `input_dim == embed_dim`, whereas it used to be `Identity`).

Please document clearly that stage-1 must be retrained before stage-2, and that old checkpoints cannot be loaded. If loading old ckpts is still required, add an explicit migration / version guard instead of failing opaquely at `load_state_dict`.

File: rlinf/models/embodiment/modules/rlt_token_transformer.py [hunk 2]
```diff
@@ -225,55 +223,88 @@ def __init__(
                 for _ in range(num_layers)
             ]
         )
-        self.output_proj = (
-            nn.Linear(self.embed_dim, self.input_dim)
-            if self.input_dim != self.embed_dim
-            else nn.Identity()
-        )
+        self.output_proj = nn.Linear(self.embed_dim, self.input_dim)
```
Comment (F9rozen) [inline rlinf/models/embodiment/modules/rlt_token_transformer.py:226]: `output_proj` is now unconditionally `nn.Linear(embed_dim, input_dim)`. Previously it was `Identity` when dims matched (the default 2048/2048 case).

Even ignoring the broader AR rewrite, this alone breaks weight compatibility for default configs. Is the Linear necessary for the AR objective, or can we keep `Identity` when `input_dim == embed_dim`?

File: rlinf/models/embodiment/modules/rlt_token_transformer.py [hunk 3]
```diff
@@ -314,15 +343,19 @@ def encode_flat(
...
-        self, rl_tokens: torch.Tensor, target_seq_len: int | None = None
+        self,
+        rl_tokens: torch.Tensor,
+        target_embeddings: torch.Tensor,
+        mask: torch.Tensor | None = None,
     ) -> torch.Tensor:
-        return self.decoder(rl_tokens, target_seq_len)
+        return self.decoder(rl_tokens, target_embeddings, mask)
 
     def reconstruct(
```
Comment (F9rozen) [inline rlinf/models/embodiment/modules/rlt_token_transformer.py:353]: `reconstruct()` now detaches the prefix and runs teacher-forced decoding. Please add unit tests covering:
1) causal mask: position `t` cannot attend to `> t`
2) loss is masked correctly when `mask` has trailing pads
3) output shape equals `prefix_embs` shape
4) gradients flow into encoder/decoder params but not into the detached teacher targets

This path is the core of the PR and currently has no direct tests.

File: rlinf/models/embodiment/modules/rlt_token_transformer.py
Comment (F9rozen) [inline rlinf/models/embodiment/modules/rlt_token_transformer.py]: Docstring still says "a small set of RL tokens", but the module is hard-coded to a single token. Please update the encoder docstring to match.

File: rlinf/algorithms/rlt/transition.py [hunk 4]
```diff
@@ -64,8 +66,26 @@ def update_rlt_transitions(
...
     cache_current: bool,
+    intervene_actions: torch.Tensor | None = None,
+    intervene_flags: torch.Tensor | None = None,
 ) -> None:
     if pending_obs[stage_id] is not None:
+        if intervene_actions is not None and intervene_flags is not None:
+            current_obs = pending_obs[stage_id]
+            ref_chunk = current_obs["ref_chunk"]
+            batch_size = ref_chunk.shape[0]
+            flags = intervene_flags.reshape(batch_size, -1, 1).to(
```
Comment (F9rozen) [inline rlinf/algorithms/rlt/transition.py:77]: `intervene_flags.reshape(batch_size, -1, 1)` assumes flags can be broadcast over the action chunk.

If `intervene_flags` is shaped `[B]` (env-level) this only patches `ref_actions[:, :1]`, not the full chunk. If it is `[B, chunk_len]` it works.

File: rlinf/algorithms/rlt/transition.py [hunk 5]
```diff
@@ -64,8 +66,26 @@ def update_rlt_transitions(
...
+                device=ref_chunk.device, dtype=torch.bool
+            )
+            human_actions = intervene_actions.reshape(batch_size, flags.shape[1], -1)
+            action_dim = human_actions.shape[-1]
+            ref_actions = ref_chunk.reshape(batch_size, -1, action_dim).clone()
+            ref_actions[:, : flags.shape[1]] = torch.where(
```
Comment (F9rozen) [inline rlinf/algorithms/rlt/transition.py:83]: `torch.where(flags, human_actions, ref_actions[:, : flags.shape[1]])` silently truncates when `flags.shape[1] != ref_chunk` horizon.

If a mismatch is unexpected, please raise; if partial override is intentional (e.g. only executed sub-steps), add a comment explaining why truncating the leading axis is correct.

**Comment by F9rozen** [comment]
Please fix the title typo ("upadte" → "update") and fill in the PR template:
- why AR reconstruction replaces the old decoder
- why single RL token is now mandatory
- checkpoint compatibility (stage-1 retrain required?)
- how intervention→ref_chunk was tested (ManiSkill expert takeover and/or real-world teleop)

**Review by guozhen1997 — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]
